### PR TITLE
License recycle upon host join modification

### PR DESCRIPTION
**Pull request by mrvinceo**
Modification to the License recycling function, to allow the licence to also be recycled when the host joins the meeting, removing risk of license being taken between creation of event and the event start time. Config option also added, defaulting to off.

### CODE REVIEW

File: classes/webservice.php [hunk 1]
```diff
@@ -444,27 +444,40 @@ protected function _database_to_api($zoom) {
...
      *
-     * @param stdClass $zoom The meeting to create.
-     * @return stdClass The call response.
+     * @param stdClass $zoom The meeting to update.
+     * @return void
      */
-    public function create_meeting($zoom) {
-        // Checks whether we need to recycle licenses and acts accordingly.
+    public function provide_license($zoom) {
+    // Checks whether we need to recycle licenses and acts accordingly.
```
Comment (jrchamp) [inline classes/webservice.php:455]:
```suggestion
        // Checks whether we need to recycle licenses and acts accordingly.
```

File: classes/webservice.php [hunk 2]
```diff
@@ -444,27 +444,40 @@ protected function _database_to_api($zoom) {
...
         if ($this->recyclelicenses && $this->_make_call("users/$zoom->host_id")->type == ZOOM_USER_TYPE_BASIC) {
             if ($this->_paid_user_limit_reached()) {
-                $leastrecentlyactivepaiduserid = $this->_get_least_recently_active_paid_user_id();
-                // Changes least_recently_active_user to a basic user so we can use their license.
+            $leastrecentlyactivepaiduserid = $this->_get_least_recently_active_paid_user_id();
```
Comment (jrchamp) [inline classes/webservice.php:458]:
```suggestion
                $leastrecentlyactivepaiduserid = $this->_get_least_recently_active_paid_user_id();
```

File: classes/webservice.php [hunk 3]
```diff
@@ -444,27 +444,40 @@ protected function _database_to_api($zoom) {
...
+            // Changes least_recently_active_user to a basic user so we can use their license.
```
Comment (jrchamp) [inline classes/webservice.php:459]:
```suggestion
                // Changes least_recently_active_user to a basic user so we can use their license.
```

File: classes/webservice.php [hunk 4]
```diff
@@ -444,27 +444,40 @@ protected function _database_to_api($zoom) {
...
                 $this->_make_call("users/$leastrecentlyactivepaiduserid", array('type' => ZOOM_USER_TYPE_BASIC), 'patch');
-            }
+        }
```
Comment (jrchamp) [inline classes/webservice.php:461]:
```suggestion
            }
```

File: classes/webservice.php [hunk 5]
```diff
@@ -444,27 +444,40 @@ protected function _database_to_api($zoom) {
...
             // Changes current user to pro so they can make a meeting.
             $this->_make_call("users/$zoom->host_id", array('type' => ZOOM_USER_TYPE_PRO), 'patch');
-        }
-
+            }
```
Comment (jrchamp) [inline classes/webservice.php:464]:
```suggestion
        }
```

File: classes/webservice.php [hunk 6]
```diff
@@ -444,27 +444,40 @@ protected function _database_to_api($zoom) {
...
+     * @param stdClass $zoom The meeting to create.
+     * @return stdClass The call response.
+     */
+    public function create_meeting($zoom) {
+    // Including the external provide_license method, previously part of this method. 
+        $this->provide_license($zoom);
         $url = "users/$zoom->host_id/" . ($zoom->webinar ? 'webinars' : 'meetings');
         return $this->_make_call($url, $this->_database_to_api($zoom), 'post');
     }
+
```
Comment (jrchamp) [inline classes/webservice.php:480]:
```suggestion
```

File: classes/webservice.php [hunk 7]
```diff
@@ -444,27 +444,40 @@ protected function _database_to_api($zoom) {
...
+    
+    /**
+     * Create a meeting/webinar on Zoom.
+     * Take a $zoom object as returned from the Moodle form and respond with an object that can be saved to the database.
+     *
+     * @param stdClass $zoom The meeting to create.
+     * @return stdClass The call response.
+     */
+    public function create_meeting($zoom) {
+    // Including the external provide_license method, previously part of this method.
```
Comment (jrchamp) [inline classes/webservice.php:475]:
```suggestion
        // Including the external provide_license method, previously part of this method. 
```

File: loadmeeting.php [hunk 8]
```diff
@@ -72,4 +72,14 @@
 // Record user's clicking join.
 \mod_zoom\event\join_meeting_button_clicked::create(array('context' => $context, 'objectid' => $zoom->id, 'other' =>
         array('cmid' => $id, 'meetingid' => (int) $zoom->meeting_id, 'userishost' => $userishost)))->trigger();
+
+// Upgrade host upon user joining meeting, if host if not Licensed [./classes/webservice.php]
+$config = get_config('mod_zoom');
+if (!empty($config->recycleonjoin)) {
+    $service = new mod_zoom_webservice();
+    if ($userishost){
```
Comment (jrchamp) [inline loadmeeting.php:80]:
```suggestion
    if ($userishost) {
```

File: classes/webservice.php [hunk 9]
```diff
@@ -444,27 +444,40 @@ protected function _database_to_api($zoom) {
     }
 
     /**
-     * Create a meeting/webinar on Zoom.
-     * Take a $zoom object as returned from the Moodle form and respond with an object that can be saved to the database.
+     * Extracted from create_meeting method.
```
Comment (jrchamp) [inline classes/webservice.php:447]: Optional suggestion, but this detail is not important.
```suggestion
```

File: classes/webservice.php [hunk 10]
```diff
@@ -444,27 +444,40 @@ protected function _database_to_api($zoom) {
...
+     * To be called from elsewhere to provide host with license
```
Comment (jrchamp) [inline classes/webservice.php:448]:
```suggestion
     * Provide the host with a license if needed and recycling is enabled.
```

File: classes/webservice.php [hunk 11]
```diff
@@ -444,27 +444,40 @@ protected function _database_to_api($zoom) {
...
+     * following creation of the meeting event.
```
Comment (jrchamp) [inline classes/webservice.php:449]: Optional suggestion, but this detail is inaccurate now that it happens on join too and the "when" it happens is unnecessary.
```suggestion
```

File: classes/webservice.php [hunk 12]
```diff
@@ -444,27 +444,40 @@ protected function _database_to_api($zoom) {
...
      *
-     * @param stdClass $zoom The meeting to create.
-     * @return stdClass The call response.
+     * @param stdClass $zoom The meeting to update.
+     * @return void
      */
-    public function create_meeting($zoom) {
-        // Checks whether we need to recycle licenses and acts accordingly.
+    public function provide_license($zoom) {
```
Comment (jrchamp) [inline classes/webservice.php:454]: The only useful part of the Zoom meeting object here is the `host_id`. This function should probably be changed to take a host_id directly. That would involve updating this newly added function's signature, the two uses of this function (that are both part of this PR), and the PHPDoc for this function.

**Comment by rlorenzo** [comment]
@mrvinceo Does this have any dependency on redefinelicenses being enabled?

**Comment by rlorenzo** [comment]
@mrvinceo @jrchamp Is this ready to go? I cannot really test it because this is not a feature we would want turned on for our Zoom (we cannot unassign/assign licenses).

If it works for you and you have it in production let me know.

File: classes/webservice.php (hunk 12 above)
Comment (mrvinceo) [reply classes/webservice.php:452]: Many thanks for all the suggestions Jonathan; this is my first foray back into php, and code in general, after about five years, and it shows! How would I go about passing the host_id directly; I assumed I'd still need  $zoom in order to get the host_id for the meeting? I'm just getting into OOP so my knowledge is patchy to say the least. Many thanks again.

Comment (jrchamp) [reply classes/webservice.php:452]: I created a PR that should address this: https://github.com/mrvinceo/moodle-mod_zoom/pull/1

**Comment by jrchamp** [comment]
@rlorenzo Should be good. @mrvinceo confirmed that it's working.

**Pull request merged**